packages/coreutils/src/restorablepool.ts

```diff
@@ -0,0 +1,378 @@
+// Copyright (c) Jupyter Development Team.
+// Distributed under the terms of the Modified BSD License.
+
+import { PromiseDelegate } from '@phosphor/coreutils';
+
+import { IObservableDisposable } from '@phosphor/disposable';
+
+import { AttachedProperty } from '@phosphor/properties';
+
+import { ISignal, Signal } from '@phosphor/signaling';
+
+import { IObjectPool, IRestorable } from './interfaces';
+
+/**
+ * An object pool that supports restoration.
+ *
+ * @typeparam T - The type of object being tracked.
+ */
+export class RestorablePool<
```

Comment:
I just realized that I have the perfect use-case for this in my LaTeX extension. A set of things that I restore from the state database with a command upon page refreshes. I think that this will tighten up my extension code a bit, and provide a good test for this during the RC phase.

Reply:
Awesome.